File: fabtrk/fabtrk/report/ft_transaction/ft_transaction.js
// // Copyright (c) 2026, UpGo Technologies and contributors
// // For license information, please see license.txt

// frappe.query_reports["FT Transaction"] = {
// 	"filters": [

// 	]
// };




frappe.query_reports["FT Transaction"] = {
	onload(report) {
		frappe.query_report.set_filter_value("project_number", "");
		frappe.query_report.set_filter_value("stages", "");
	},

	filters: [
		// ---------------- PROJECT ----------------
		{
			fieldname: "project_number",
			label: "Project",
			fieldtype: "Link",
			options: "FT Project Stages",
			on_change() {
				frappe.query_report.set_filter_value("stages", "");
				frappe.query_report.refresh();
			}
		},

		// ---------------- STAGES ----------------
		{
			fieldname: "stages",
			label: "Stages",
			fieldtype: "Link",
			options: "FT Stages Master",
			on_change() {
				frappe.query_report.refresh();
			}
		}
	],

	// ── Click on Project or Stage header → set filter & refresh ──
	after_datatable_render(report) {
		// Project cell click → filter set karo
		$(report.wrapper)
			.off("click", ".project-link")
			.on("click", ".project-link", function () {
				let project = $(this).data("project");
				if (project) {
					frappe.query_report.set_filter_value("project_number", project);
					frappe.query_report.refresh();
				}
			});

		// Stage cell click → filter set karo
		$(report.wrapper)
			.off("click", ".stage-link")
			.on("click", ".stage-link", function () {
				let stage = $(this).data("stage");
				if (stage) {
					frappe.query_report.set_filter_value("stages", stage);
					frappe.query_report.refresh();
				}
			});
	},

	// formatter(value, row, column, data, default_formatter) {
	// 	value = default_formatter(value, row, column, data);

	// 	// TOTAL row bold styling
	// 	if (data && data.is_total) {
	// 		return `<span style="font-weight:700; background:#f0f4ff;">${value}</span>`;
	// 	}

	// 	// Project column — clickable link style
	// 	if (column.fieldname === "project_number" && data && data.project_number) {
	// 		return `<span class="project-link" data-project="${data.project_number}"
	// 					style="color:#1F4E79; font-weight:600; cursor:pointer;
	// 				">
	// 					${data.project_number}
	// 				</span>`;
	// 	}

	// 	// Stage column — clickable link style
	// 	if (column.fieldname === "stage" && data && data.stage) {
	// 		return `<span class="stage-link" data-stage="${data.stage}"
	// 					style="color:#0c5c70; font-weight:600; cursor:pointer;
	// 				">
	// 					${data.stage}
	// 				</span>`;
	// 	}

	// 	// Balance column — color coding
	// 	if (column.fieldname === "balance" && data) {
	// 		let val = parseFloat(data.balance || 0);
	// 		let color = val > 0 ? "#1a7abf" : val < 0 ? "#c00000" : "#555";
	// 		return `<span style="font-weight:700; color:${color};">${val}</span>`;
	// 	}

	// 	// Available column styling
	// 	if (column.fieldname === "available" && data) {
	// 		return `<span style="font-weight:600; color:#155724;">${value}</span>`;
	// 	}

	// 	return value;
	// }
	formatter(value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);

		// TOTAL row bold styling
		if (data && data.is_total) {
			return `<span style="font-weight:700; background:#f0f4ff;">${value}</span>`;
		}

		// Project column — clickable link style
		if (column.fieldname === "project_number" && data && data.project_number) {
			return `<span class="project-link" data-project="${data.project_number}"
                    style="cursor:pointer;">
                    ${data.project_number}
                </span>`;
		}

		// Stage column — clickable link style
		if (column.fieldname === "stage" && data && data.stage) {
			return `<span class="stage-link" data-stage="${data.stage}"
                    style="cursor:pointer;">
                    ${data.stage}
                </span>`;
		}

		return value;
	}
};

File: fabtrk/fabtrk/report/ft_transaction/ft_transaction.py
# # Copyright (c) 2026, UpGo Technologies and contributors
# # For license information, please see license.txt

# # import frappe


# def execute(filters=None):
# 	columns, data = [], []
# 	return columns, data


import frappe

def execute(filters=None):
    filters = filters or {}

    # ── Columns ──
    columns = [
        {
            "label": "Project",
            "fieldname": "project_number",
            "fieldtype": "Link",
            "options": "FT Project Stages",
            "width": 160,
        },
        {
            "label": "Stage",
            "fieldname": "stage",
            "fieldtype": "Data",
            "width": 180,
        },
        {
            "label": "Available",
            "fieldname": "available",
            "fieldtype": "Int",
            "width": 200,
            "align": "center",
        },
        {
            "label": "Addition",
            "fieldname": "addition",
            "fieldtype": "Int",
            "width": 190,
            "align": "center",
        },
        {
            "label": "Subtraction",
            "fieldname": "subtraction",
            "fieldtype": "Int",
            "width": 120,
            "align": "center",
        },
        {
            "label": "Completed",
            "fieldname": "completed",
            "fieldtype": "Int",
            "width": 190,
            "align": "center",
        },
        {
            "label": "Balance",
            "fieldname": "balance",
            "fieldtype": "Int",
            "width": 150,
            "align": "center",
        },
    ]

    # ── Conditions ──
    conditions = "WHERE 1=1"
    values = {}

    if filters.get("project_number"):
        conditions += " AND ps.name = %(project_number)s"
        values["project_number"] = filters["project_number"]

    if filters.get("stages"):
        conditions += " AND sm.name = %(stages)s"
        values["stages"] = filters["stages"]

    # ── Main Query ──
    # FT Project Stages → Child (stages list) → FT Transaction (addition, subtraction, completed)
    # Available = SUM(addition) - SUM(subtraction) - SUM(completed) so far
    # Balance   = Available - Completed (current period)
    raw_data = frappe.db.sql(f"""
        SELECT
            ps.name                         AS project_number,
            sm.name                         AS stage,
            psc.stages_key                  AS stages_key,

            -- Transactions aggregate karo per project+stage
            COALESCE(SUM(t.addition), 0)    AS addition,
            COALESCE(SUM(t.subtraction), 0) AS subtraction,
            COALESCE(SUM(t.completed), 0)   AS completed

        FROM `tabFT Project Stages` ps
        INNER JOIN `tabFT Project Stages Child` psc
            ON psc.parent = ps.name
        INNER JOIN `tabFT Stages Master` sm
            ON sm.name = psc.stages

        -- Transactions left join (ho sakta hai koi transaction na ho)
        LEFT JOIN `tabFT Transaction` t
            ON t.project_number = ps.name
            AND t.stages = psc.name

        {conditions}

        GROUP BY
            ps.name,
            sm.name,
            psc.stages_key

        ORDER BY
            ps.name ASC,
            psc.stages_key ASC
    """, values, as_dict=True)

    if not raw_data:
        return columns, []

    # ── Calculate Available & Balance ──
    # Available = cumulative: pichle saare addition - subtraction - completed
    # Balance   = Available (running total after this row's transactions)
    # Logic per project:
    #   available (start of stage) = previous balance
    #   balance   = available + addition - subtraction - completed

    data = []
    grand = {
        "addition": 0,
        "subtraction": 0,
        "completed": 0,
        "available": 0,
        "balance": 0,
    }

    # Group by project for running balance
    from collections import defaultdict
    project_rows = defaultdict(list)
    for row in raw_data:
        project_rows[row["project_number"]].append(row)

    for project, rows in project_rows.items():
        running_balance = 0  # har project ke liye fresh start

        for row in rows:
            addition    = int(row.get("addition")    or 0)
            subtraction = int(row.get("subtraction") or 0)
            completed   = int(row.get("completed")   or 0)

            available = running_balance  # is stage ke start mein kitna tha
            balance   = available + addition - subtraction - completed
            running_balance = balance   # agla stage yahan se start karega

            data.append({
                "project_number": project,
                "stage":          row["stage"],
                "available":      available,
                "addition":       addition,
                "subtraction":    subtraction,
                "completed":      completed,
                "balance":        balance,
                "is_total":       0,
            })

            grand["addition"]    += addition
            grand["subtraction"] += subtraction
            grand["completed"]   += completed
            grand["available"]   += available
            grand["balance"]     += balance

    # ── Grand Total Row ──
    data.append({
        "project_number": "TOTAL",
        "stage":          "",
        "available":      grand["available"],
        "addition":       grand["addition"],
        "subtraction":    grand["subtraction"],
        "completed":      grand["completed"],
        "balance":        grand["balance"],
        "is_total":       1,
    })

    return columns, data


